lib/CodeGen/CodeGenModule.cpp

```diff
@@ -2009,77 +2011,93 @@ bool CodeGenModule::shouldOpportunisticallyEmitVTables() {
-      // let hc_grid_launch kernel functions to pass
-      D->hasAttr<HCGridLaunchAttr>())
-    return true;
+    bool allowed_(const VarDecl* x)
+    {
+      if (!x) return true;
+      if (d_.count(x)) return d_[x];
 
-  // the remaining ones must be functions
-  // be selective about functions to pass
-
-  StringRef MangledName = CGM.getMangledName(GD);
-
-  // C++11 atomic functions are allowed to pass
-  // C++11 functional operators are allowed to pass
-  // C++11 swap functions are allowed to pass
-  // C++11 numeric limits are allowed to pass
-  // C++11 pair are allowed to pass
-  // C++11 complex are allowed to pass
-  // PSTL operators are allowed to pass
-  const CXXMethodDecl* MethodD = dyn_cast<CXXMethodDecl>(D);
-  if (MethodD) {
-    StringRef ClassName = MethodD->getParent()->getName();
-    if (ClassName.find("__atomic_base") != StringRef::npos ||
-        ClassName.find("plus") != StringRef::npos ||
-        ClassName.find("logical_or") != StringRef::npos ||
-        ClassName.find("logical_and") != StringRef::npos ||
-        ClassName.find("unique_ptr") != StringRef::npos ||
-        ClassName.find("compressed_pair") != StringRef::npos ||
-        ClassName.find("numeric_limits") != StringRef::npos ||
-        ClassName.find("pair") != StringRef::npos ||
-        ClassName.find("complex") != StringRef::npos ||
-        MangledName.find("experimental8parallel") != StringRef::npos) {
-      return true;
+      bool r = true;
+
+      if (!x->hasAttr<HCCTileStaticAttr>() &&
+          (x->isStaticLocal() ||
+          x->hasExternalStorage() ||
+          x->hasGlobalStorage() ||
+          x->isExceptionVariable()))  {
+            r = false;
+      }
+
+      d_[x] = r;
+
+      return r;
     }
-  }
 
-  // C++11 swap/move functions are allowed to pass
-  // C++11 memory_order modifiers are allowed to pass
-  // C++11 complex operators are allowed to pass
-  // certain C++11 math functions are allowed to pass
-  // Eigen internal functions are allowed to pass
-  const FunctionDecl* FuncD = dyn_cast<FunctionDecl>(D);
-  if (FuncD) {
-    if (MangledName.find("4swap") != StringRef::npos ||
-        MangledName.find("4move") != StringRef::npos ||
-        MangledName.find("16grid_launch_parm10KernelArgsIT2_EENKUlvE_clEv") != StringRef::npos ||
-        MangledName.find("St12memory_order") != StringRef::npos ||
-        MangledName.find("Kokkos4Impl") != StringRef::npos ||
-        MangledName.find("St16initializer_list") != StringRef::npos ||
-        MangledName.find("St7complex") != StringRef::npos ||
-        MangledName.find("5roundf") != StringRef::npos ||
-        MangledName.find("Eigen8internal") != StringRef::npos) {
-      return true;
+    bool allowed_(const FunctionDecl* x)
+    {
+      if (!x) return true;
+      if (d_.count(x)) return d_[x];
+
+      bool r = true;
+
+      if (x->isVariadic()) r = false;
+      if (x->isPure() || x->isVirtualAsWritten()) r = false;
```

Comment:
@scchan this is to check if the function is a pure virtual function. I think it's fine here.